### PR TITLE
[Merged by Bors] - feat: Port/Combinatorics.Derangements.Basic

### PR DESCRIPTION
**Pull request by arienmalec**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

Previously had dependencies on a `Type`/`Sort` issue between Mathlib3/4

Plan:

- [x] Fix current issue (proof that `DecidableEq a → DecidablePred (fun (x: α) => x ∈ ({a}ᶜ : Set α))`
- [x] Wait on https://github.com/leanprover-community/mathlib/pull/18543 to turn into a PR on the mathlib4 side
- [x] Take a dependency on that PR
- [x] Re-port `Type`/`Sort` dependent changes

### CODE REVIEW

**Comment by arienmalec** [comment]
I have this down to a single issue in `atMostOneFixedPointEquivSumDerangements`

The original issue was instance timeouts, which I fixed by getting the value of the type hole on the Lean 3 side.

I now have a partly sorried proof that relies on a proof or declaration that `DecidablePred (fun (x: α) => x ∈ ({a}ᶜ : Set α))`

File: Mathlib/Logic/Equiv/Basic.lean [hunk 1]
```diff
@@ -1165,7 +1165,7 @@ def subtypeSubtypeEquivSubtypeExists (p : α → Prop) (q : Subtype p → Prop)
 
 /-- A subtype of a subtype is equivalent to the subtype of elements satisfying both predicates. -/
 @[simps!]
-def subtypeSubtypeEquivSubtypeInter (p q : α → Prop) :
+def subtypeSubtypeEquivSubtypeInter {α : Type u} (p q : α → Prop) :
```
Comment (eric-wieser) [inline Mathlib/Logic/Equiv/Basic.lean:1168]: [#18543](https://github.com/leanprover-community/mathlib/pull/18543) contains a backport of the original change, as opposed to this PR which reverts the change.

File: Mathlib/Combinatorics/Derangements/Basic.lean [hunk 2]
```diff
@@ -16,193 +16,210 @@ import Mathlib.Logic.Equiv.Option
...
+In order to prove the above, we also prove some results about the effect of `Equiv.removeNone`
+on derangements: `RemoveNone.fiberNone` and `RemoveNone.fiber_some`.
 -/
 
 
 open Equiv Function
 
 /-- A permutation is a derangement if it has no fixed points. -/
-def derangements (α : Type _) : Set (Perm α) :=
+def Derangements (α : Type _) : Set (Perm α) :=
```
Comment (eric-wieser) [inline Mathlib/Combinatorics/Derangements/Basic.lean:36]: This should remain lowercase

Comment (arienmalec) [reply Mathlib/Combinatorics/Derangements/Basic.lean:36]: I upcased by analogy with `PNat`, etc, contra mathport. Brought to Zulip for discussion.

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**